src/box/lua/slab.cc

```diff
@@ -265,6 +275,126 @@ lbox_slab_check(MAYBE_UNUSED struct lua_State *L)
 	return 0;
 }
 
+#if defined(HAVE_MALLOC_INFO)
+/*
+ * Decodes and returns the XML document returned by malloc_info() as is.
+ * This is an internal function that isn't supposed to be called by users,
+ * but it may be useful for developers.
+ */
+static int
+lbox_malloc_internal_info(struct lua_State *L)
+{
+	char *buf;
+	size_t buf_size;
+	FILE *fp = open_memstream(&buf, &buf_size);
+	if (fp == NULL) {
+		diag_set(SystemError, "failed to open memory stream");
+		return luaT_error(L);
+	}
+	if (malloc_info(0, fp) != 0) {
+		diag_set(SystemError, "failed to get malloc info");
+		return luaT_error(L);
+	}
+	fclose(fp);
+	lua_pushlstring(L, buf, buf_size);
+	free(buf);
+	return luaT_xml_decode(L);
+}
+
+/*
+ * Returns the malloc memory usage information in a table
+ *
+ *   {
+ *     size = <total allocated>,
+ *     used = <actually used>,
+ *   }
+ *
+ * (all numbers are in bytes).
+ */
+static int
+lbox_malloc_info(struct lua_State *L)
+{
+	lbox_malloc_internal_info(L);
+	/*
+	 * The XML document name returned by malloc_info is expected to be
+	 * "malloc" so the document content should be in malloc[1].
+	 */
+	assert(lua_istable(L, -1));
+	lua_getfield(L, -1, "malloc");
+	if (lua_istable(L, -1))
+		lua_rawgeti(L, -1, 1);
+	/*
+	 * First, check the malloc_info version. The only known version is 1.
+	 * It's pointless to proceed if the version is different.
+	 */
+	int version = 0;
+	if (lua_istable(L, -1)) {
+		lua_getfield(L, -1, "version");
+		version = lua_tonumber(L, -1);
+		lua_pop(L, 1);
+	}
+	if (version != 1)
+		return luaL_error(L, "unknown malloc info format");
+	/*
+	 * Extract the size of used memory. Even though the document version is
+	 * valid, we still need to be careful accessing it.
+	 */
+	uint64_t total = 0;
+	uint64_t available = 0;
+	assert(lua_istable(L, -1));
+	lua_getfield(L, -1, "system");
+	if (lua_istable(L, -1)) {
+		lua_pushnil(L);
+		for (; lua_next(L, -2) != 0; lua_pop(L, 1)) {
+			if (!lua_istable(L, -1))
+				continue;
```

Comment:
This condition is not clear at glance. Isn't a table supposed by the format that the internal XML parser returns? I would acquire some brief notes.

Reply:
Both XML attributes and sub-elements are stored at the same level. This condition skips XML attributes. I added a comment clarifying this.

<details><summary>diff</summary>

```diff
diff --git a/src/box/lua/slab.cc b/src/box/lua/slab.cc
index fc833516f9b4..0006a59f02d1 100644
--- a/src/box/lua/slab.cc
+++ b/src/box/lua/slab.cc
@@ -345,33 +358,31 @@ lbox_malloc_info(struct lua_State *L)
 		lua_pushnil(L);
 		for (; lua_next(L, -2) != 0; lua_pop(L, 1)) {
 			if (!lua_istable(L, -1))
-				continue;
+				continue; /* skip XML attributes */
 			lua_getfield(L, -1, "type");
 			const char *key = lua_tostring(L, -1);
 			lua_pop(L, 1);
@@ -367,7 +378,7 @@ lbox_malloc_info(struct lua_State *L)
 		lua_pushnil(L);
 		for (; lua_next(L, -2) != 0; lua_pop(L, 1)) {
 			if (!lua_istable(L, -1))
-				continue;
+				continue; /* skip XML attributes */
 			lua_getfield(L, -1, "type");
 			const char *key = lua_tostring(L, -1);
 			lua_pop(L, 1);
```

</details>